### PR TITLE
Enumerations, SchemaDefRequirement and fix for nested Inputbindings

### PR DESCRIPTION
**Pull request by danbills**
Borne of a requirement on GDC Dnaseq using Enumerations.

Represents a spike on Test \#3 which was the only one that tested Enumerations at the time of this writing.

A list of features/fixes

* implementation of SchemaDefRequirement.  This produced a lot of noise as any time `Myriad` types were used this must be provided for reference.
* nested input bindings sort properly. 
* impl of EnumSchema for input & output
* fix for coproduct of types -> command parts ( previously was using values from fields that matched any field of any coproduct type.  this fix regards each type holistically and thus rejects unless all fields match)
* Introduced a `RequirementsAndHints` class to handle common operations on `List[Requirement]`.

## ~TODO~

~I broke conf test \#71 and centaur test dealing w/ secondary files.  I expect fixes to be minor.~

### CODE REVIEW

File: cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala [hunk 1]
```diff
@@ -72,18 +88,20 @@ object MyriadInputTypeToSortedCommandParts extends Poly1 {
...
 
   // Primitive type: we just need to create a command part from the binding if there's one here.
   implicit def ct: Aux[CwlType, CommandPartBuilder] = {
     at {
       cwlType => {
-        (inputBinding, womValue, key, hasShellCommandRequirement, expressionLib) => {
-          cwlType match {
-            case CwlType.Null => CommandPartsList.empty
-            case _ => inputBinding.toList.map(_.toCommandPart(key, womValue, hasShellCommandRequirement, expressionLib))
+        case (_, WomOptionalValue(_, None), _, _, _, _) => CommandPartsList.empty.some
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala]:
```
scala> import cats.syntax.option._
import cats.syntax.option._

scala> val s: String = null
s: String = null

scala> s.some
res0: Option[String] = Some(null)

scala> Option(s)
res1: Option[String] = None
```

😞

File: cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala [hunk 2]
```diff
@@ -94,111 +112,144 @@ object MyriadInputInnerTypeToSortedCommandParts extends Poly1 {
...
+      case (None, _, _, _, _, _) if !irs.fields.exists(_.exists(_.inputBinding.isDefined)) => CommandPartsList.empty.some
 
-      case (inputBinding, objectLike: WomObjectLike, sortingKey, hasShellCommandRequirement, expressionLib) =>
+      case (inputBinding, objectLike: WomObjectLike, sortingKey, hasShellCommandRequirement, expressionLib, schemaDefRequirement) =>
         // If there's an input binding, make a SortKeyAndCommandPart for it
-        val fromInputBinding: Option[SortKeyAndCommandPart] =
+        val sortingKeyFromInputBindingFromInputParameter: Option[SortKeyAndCommandPart] =
           inputBinding.map(_.toCommandPart(sortingKey, objectLike, hasShellCommandRequirement, expressionLib))
 
         // Go over the fields an fold over their type
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala]: and fold

File: cwl/src/main/scala/cwl/internal/CommandPartSortingAlgorithm.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package cwl.internal
+
+import cats.data.Kleisli._
+import cats.data.ReaderT
+import cats.data.Validated.{Invalid, Valid, _}
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/internal/CommandPartSortingAlgorithm.scala]: aka _

File: centaur/src/main/resources/standardTestCases/cwl_secondary_files/cwl_secondary_files.cwl [hunk 4]
```diff
@@ -29,6 +29,7 @@ $graph:
       secondaryFiles: [.also]
       type:
         type: record
+        name: foo
```
[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/cwl_secondary_files/cwl_secondary_files.cwl]: Hm, I guess this should be optional in `InputRecordSchema`.  If you look there you'll see I added a `name` field.  This is contentious as `name` field appeared in conformance tests but not the spec!  I'm off to add a comment in the code.

[the remark this answers is not in the document]
Comment (danbills) [reply centaur/src/main/resources/standardTestCases/cwl_secondary_files/cwl_secondary_files.cwl]: So it is non-optional in v1.1 so I've kept it as such. Thus this data remains to conform to the (pseudo-)spec.

File: cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala (hunk 1 above)
Comment (danbills) [reply cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala]: I get your point, but this instance is guaranteed non-null.  I'm working on a `.option` PR for cats for general usage.

Comment (danbills) [reply cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala]: https://github.com/typelevel/cats/pull/2235

File: cwl/src/main/scala/cwl/CommandOutputBinding.scala [hunk 5]
```diff
@@ -220,7 +220,13 @@ object CommandOutputBinding {
         val globs = List(cwlPath)
         globs traverse loadFileWithContents(ioFunctionSet, commandOutputBinding)
       case WomMaybePopulatedFileType =>
-        val globs = Await.result(ioFunctionSet.glob(cwlPath), Duration.Inf)
+        val globs: Seq[String] = Await.result(ioFunctionSet.glob(cwlPath), Duration.Inf).
+
+        //TODO: HACK ALERT - DB: I am starting on ticket https://github.com/broadinstitute/cromwell/issues/3472 which will redeem me of this mortal sin.
+          filterNot{s =>
+            s.endsWith("rc.tmp") || s.endsWith("docker_cid") || s.endsWith("script") || s.endsWith("script.background") || s.endsWith("script.submit") || s.endsWith("stderr") || s.endsWith("stderr.background") || s.endsWith("stdout") || s.endsWith("stdout.background")
```
Comment (danbills) [inline cwl/src/main/scala/cwl/CommandOutputBinding.scala]: I did in fact push this code on purpose to see if it causes damage or even passes the detritus test.  It got me over the hump on GDC bamfastq_align. I am starting ticket #3092 to provide an actual solution.

**Pull request merged**